AdvLoggerPkg/Library/AdvancedLoggerLib/Pei/GoogleTest/AdvancedLoggerPeiLibGoogleTest.cpp

```diff
@@ -0,0 +1,225 @@
+  EXPECT_CALL (
+    AdvancedLoggerPpiMock,
+    gAL_AdvancedLoggerWritePpi (
+      Eq (DebugLevel),
+      BufferEq (Buffer, NumberOfBytes),
+      Eq (NumberOfBytes)
+      )
+    )
+    .WillOnce (
+       Return ()
+       );
+
+  AdvancedLoggerWrite (DebugLevel, Buffer, NumberOfBytes);
+}
+
+/* Passing an invalid buffer - should be caught/handled by the protocol */
+TEST_F (AdvancedLoggerWriteTestPei, AdvLoggerWriteInvalidBuffer) {
+  Buffer = nullptr;
+  UINTN  numBytesZero = 0;
+
+  // Expect the call to LocatePpi to return success, "found" AdvancedLoggerPpi (mocked)
+  EXPECT_CALL (
+    PeiServicesMock,
+    PeiServicesLocatePpi (
+      BufferEq (&gAdvancedLoggerPpiGuid, sizeof (EFI_GUID)),
+      Eq (0),
+      IsNull (),
+      NotNull ()
+      )
+    )
+    .WillOnce (
+       DoAll (
+         SetArgPointee<3>(ByRef (gALPpi)),
+         Return (EFI_SUCCESS)
+         )
+       );
+
+  // Expect the call to AdvancedLoggerWritePpi
+  EXPECT_CALL (
+    AdvancedLoggerPpiMock,
+    gAL_AdvancedLoggerWritePpi (
+      Eq (DebugLevel),
+      BufferEq (Buffer, numBytesZero),
+      Eq (NumberOfBytes)
+      )
+    )
+    .WillOnce (
+       Return ()
+       );
+
+  AdvancedLoggerWrite (DebugLevel, Buffer, NumberOfBytes);
+}
+
+/* Attempting to write 0 bytes - should be caught/handled by the protocol */
+TEST_F (AdvancedLoggerWriteTestPei, AdvLoggerWriteZeroBytes) {
+  UINTN  NumberOfBytesZero = 0;
+
+  // Expect the call to LocatePpi to return success, "found" AdvancedLoggerPpi (mocked)
+  EXPECT_CALL (
+    PeiServicesMock,
+    PeiServicesLocatePpi (
+      BufferEq (&gAdvancedLoggerPpiGuid, sizeof (EFI_GUID)),
+      Eq (0),
+      IsNull (),
+      NotNull ()
+      )
+    )
+    .WillOnce (
+       DoAll (
+         SetArgPointee<3>(ByRef (gALPpi)),
+         Return (EFI_SUCCESS)
+         )
+       );
+
+  // Expect the call to AdvancedLoggerWritePpi
+  EXPECT_CALL (
+    AdvancedLoggerPpiMock,
+    gAL_AdvancedLoggerWritePpi (
+      Eq (DebugLevel),
+      BufferEq (Buffer, NumberOfBytes),
+      Eq (NumberOfBytesZero)
+      )
+    )
+    .WillOnce (
+       Return ()
+       );
+
+  AdvancedLoggerWrite (DebugLevel, Buffer, NumberOfBytesZero);
+}
+
+/* Passing a mismatched signature. Asserts are disabled so execution will continue */
+TEST_F (AdvancedLoggerWriteTestPei, AdvLoggerWriteFailMismatchedSignature) {
+  // Expect the call to LocatePpi to return success, "found" AdvancedLoggerPpi (mocked)
+  gALPpi->Signature = SIGNATURE_32 ('T', 'E', 'S', 'T');
+  gALPpi->Version   = ADVANCED_LOGGER_PPI_VERSION + 1;
+
+  EXPECT_CALL (
+    PeiServicesMock,
+    PeiServicesLocatePpi (
+      BufferEq (&gAdvancedLoggerPpiGuid, sizeof (EFI_GUID)),
+      Eq (0),
+      IsNull (),
+      NotNull ()
+      )
+    )
+    .WillOnce (
+       DoAll (
+         SetArgPointee<3>(ByRef (gALPpi)),
+         Return (EFI_SUCCESS)
+         )
+       );
+
+  // TODO - why do we still expect a call to AdvancedLoggerWritePpi? shouldn't we exit?

```

Comment:
@apop5 Is this expected behavior?

Reply:
![image](https://github.com/user-attachments/assets/ea1d5cee-a5ed-422d-8e84-97f86a68a6e5)

With asserts disabled, it will still call. I agree with your question, and I think that if the singatures are not valid, we should just exit without writing

@kuqin12 @os-d Do you have any problem with the system not writing when the signatures are bad?

Reply:
No, I agree if the signature is bad, we shouldn't write there, we would be corrupting memory.

Reply:
Changed in this PR: https://github.com/microsoft/mu_plus/pull/565